### ISSUE TITLE
non-zero skip_quoted_offset does not properly skip quoted text

### ISSUE THREAD

**Issue by ebblake** [neomutt/neomutt#2965]
## Expected Behaviour
Using 'S' (&lt;skip-quoted&gt;) to skip to new content within a message should always locate the next unquoted line, then adjust the portion of the message displayed to include that line plus any context specified in skip_quoted_offset lines.

Skip to unquoted text, but with context still visible, all without landing on a screenful of only quoted text.

## Actual Behaviour

When skip_quoted_offset is 0, or when new content is already visible in the screen, &lt;skip-quoted&gt; is correctly scrolling the message such that the next new content plus that many context lines is moved to the top of the message window.  But when skip_quoted_text is nonzero and there is no visible new content on the screen, &lt;skip-quoted&gt; is behaving more like &lt;next-page&gt; and only scrolling through one screen-ful of information rather than all the way to the next unquoted line, requiring a lot more 'S' presses to go through a lengthy message that is mostly quoted text.

Using a non-zero skip_quoted_offset is not actually skipping quoted text!

## Steps to Reproduce

See the attached screencast (asciinema play pager_S_bug.asciinema), where all I did was open a message in the pager, then hit 'S' until the error message of "no more quoted text".  I used a single message to demo; note that the first time through (neomutt's default of skip_quoted_text=0), the message jumps to the first body line (10%) then the the first new content (96%), before a few more partial-screen scrolls to get to the final end.  After changing skip_quoted_text to 3, and repeating, you now see that I had to scroll through the entire message (from 10% to 14% to 18% ...).

## How often does this happen?

- Always

I'm attaching both the screencast and the sample mail I tested on, originally at https://lists.gnu.org/archive/html/qemu-devel/2021-06/msg04161.html

## When did it start to happen?

Don't know how long it has been present (I just started using neomutt this year)

## NeoMutt Version

```
NeoMutt 20210205
Copyright (C) 1996-2020 Michael R. Elkins and others.
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Linux 5.12.10-300.fc34.x86_64 (x86_64)
ncurses: ncurses 6.2.20200222 (compiled with 6.2.20200222)
libidn2: 2.3.1 (compiled with 2.3.0)
GPGME: 1.15.1
GnuTLS: 3.6.15
libnotmuch: 5.3.0
PCRE2: 10.36 2020-12-04
storage: tokyocabinet, lmdb
compression: lz4, zlib, zstd

Configure options: --sysconfdir=/etc --full-doc SENDMAIL=/usr/sbin/sendmail ISPELL=/usr/bin/hunspell --autocrypt --gnutls --gpgme --gss --disable-idn --idn2 --lmdb --lua --lz4 --notmuch --pcre2 --sasl --tokyocabinet --zlib --zstd

Compilation CFLAGS: -g -O2 -std=c99 -fno-delete-null-pointer-checks -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/usr/include -DNCURSES_WIDECHAR

Default options:
  +attach_headers_color +compose_to_sender +compress +cond_date +debug 
  +encrypt_to_self +forgotten_attachments +forwref +ifdef +imap +index_color 
  +initials +limit_current_thread +multiple_fcc +nested_if +new_mail +nntp +pop 
  +progress +quasi_delete +regcomp +reply_with_xorig +sensible_browser +sidebar 
  +skip_quoted +smtp +status_color +timeout +tls_sni +trash 

Compile options:
  +autocrypt +bkgdset +color +curs_set +fcntl -flock -fmemopen +futimens 
  +getaddrinfo +gnutls +gpgme +gss +hcache -homespool +idn +inotify 
  -locales_hack +lua +meta -mixmaster +nls +notmuch -openssl +pcre2 +pgp +sasl 
  +smime +sqlite +start_color +sun_attachment +typeahead 
MAILPATH="/var/mail"
PKGDATADIR="/usr/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/etc"

To learn more about NeoMutt, visit: https://neomutt.org
If you find a bug in NeoMutt, please raise an issue at:
    https://github.com/neomutt/neomutt/issues
or send an email to: <neomutt-devel@neomutt.org>

```

## Extra Info

* Operating System and its version
Fedora 34
* Were you using multiple copies of NeoMutt at once?
no
* Were you using 'screen' or 'tmux'?
no
* Is your email local (maildir) or remote (IMAP)?
maildir in the screencast (notmuch when I first encountered it)

FWIW, I first experimented with skip_quoted_offset because I noticed that upstream mutt is considering adding the same feature under a different name pager_skip_quoted_context (just now, nearly 5 years after neomutt first had this feature): https://marc.info/?l=mutt-dev&m=162372614300939&w=2.  Obviously, once upstream mutt has a working feature, we'll want to alias our old name to match theirs, and make sure the behavior is sane between the two implementations.
[pager_S_bug.asciinema](https://github.com/neomutt/neomutt/files/6664472/pager_S_bug.txt)
[7381c6b6-ce9f-11eb-908a-8c1645a91023:2,S](https://github.com/neomutt/neomutt/files/6664476/7381c6b6-ce9f-11eb-908a-8c1645a91023.2.S.txt)

**Comment by ebblake** [comment]
Instead of using
```
set skip_quoted_offset=3
```
I can mostly work around things by
```
set skip_quoted_offset = 0
macro pager S '<skip-quoted><previous-line><previous-line><previous-line>'
```
but that's a pain to type, and not as flexible for changing the value on the fly

**Comment by ebblake** [comment]
Here's what I'm playing with while comparing mutt commit 9155c6576d with neomutt.
[1624041340.R8270633834942109656.blue:2,S.txt](https://github.com/neomutt/neomutt/files/6679053/1624041340.R8270633834942109656.blue.2.S.txt)
[1624041340.R8270633834942109656.blue:2,S.txt](https://github.com/neomutt/neomutt/files/6679054/1624041340.R8270633834942109656.blue.2.S.txt)
I also quickly learned that mutt lacks `<skip-headers>' (H), and merely treats ALL headers as non-quoted text, compared to neomutt's `<skip-quoted>` S treating the first line of the body as a valid skip point if it is not quoted (we may want to add some heuristics to treat "On DATE, so-and-so wrote:" attribution lines as quoted text...)

**Comment by ebblake** [comment]
Repeating those files inline:
```
Date: Fri, 18 Jun 2021 13:35:40 -0500
From: alice@example.com
To: bob@example.com
Subject: testing skip-quoted
MIME-Version: 1.0
Content-Type: text/plain; charset=us-ascii
Content-Disposition: inline
Message-ID: <1234@example.com>
X-Mutt-Fcc: ~/sent

expected line 1

lots
of
random
filler,
all
of
which
forms
new
content.

skip-quoted
should
never
jump
to
the
middle
of
a
long
sequence
of
new
content.

> Here we have
> a few lines
> of quoted
> text

expected line 2

more
random
text

> just one line of quoted text

expected line 3

> Now, for a really long line
> Suspendisse urna, enim, elementum magnis. Montes viverra ve, bibendum quisque egestas. Sollicitudin feugiat, venenatis dui feugiat a, rutrum cras, vulputate cursus sit tellus. Proin eros a lacus. Et, suspendisse urna, hymenaeos nec, eros. Cursus duis pede vel nunc eleifend. Lacus. Id aptent auctor, amet, duis conubia pretium turpis. Est et elementum euismod, aptent eros imperdiet aptent nascetur quis, per. Quam erat dolor tortor at erat et dolor urna nibh sed. Tempor felis pede a diam lorem placerat. Dictum, magnis ac primis ridiculus leo, dapibus tempus pede pede eni.

expected line 4

nothing to see beyond this point
```
and
```
Date: Fri, 18 Jun 2021 13:38:29 -0500
From: bob@example.com
To: alice@example.com
Subject: Re: testing skip-quoted
MIME-Version: 1.0
Content-Type: text/plain; charset=us-ascii
Content-Disposition: inline
X-Mutt-Fcc: ~/sent
Message-ID: <2345@example.com>
In-Reply-To: <1234@example.com>
Received-SPF: pass (example.com: domain of list-bounces+bob=example.com@example.com designates 127.0.0.1 as permitted sender) client-ip=127.0.0.1;
Delivered-To: alice@example.com
Received: from relay.example.com (relay.example.com [127.0.0.1])
	by smtp.example.com (Postfix) with ESMTPS id C4B47216E1F7
	for <alice@example.com>; Wed, 16 Jun 2021 12:27:26 +0000 (UTC)
Received: from relay.example.com (relay.example.com [127.0.0.1])
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by relay.example.com (Postfix) with ESMTPS id AAE378007B1
	for <alice@example.com>; Wed, 16 Jun 2021 12:27:26 +0000 (UTC)
Received: from lists.example.com (lists.example.com [127.0.0.1]) (Using TLS) by
 relay.example.com with ESMTP id us-mta-470-1NEeCtqgPNWisAavKGsCEQ-1; Wed,
 16 Jun 2021 08:27:24 -0400
X-MC-Unique: 1NEeCtqgPNWisAavKGsCEQ-1
Received: from localhost ([::1]:59082 helo=another.example.com)
	by another.example.com with esmtp (Exim 4.90_1)
	(envelope-from <list-bounces+bob=example.com@example.com>)
	id 1ltUdg-0006wY-CU
	for alice@example.com; Wed, 16 Jun 2021 08:27:24 -0400
Received: from another.example.com ([2001:470:142:3::10]:60528)
 by next-step.example.com with esmtps (TLS1.2:ECDHE_RSA_AES_256_GCM_SHA384:256)
 (Exim 4.90_1) (envelope-from <bob@example.com>)
 id 1ltUct-0005jI-4J; Wed, 16 Jun 2021 08:26:35 -0400
Received: from next-step.example.com ([::1234]:39729)
 by made-up.example.com with esmtps (TLS1.2:ECDHE_RSA_AES_128_GCM_SHA256:128)
 (Exim 4.90_1) (envelope-from <bob@example.com>)
 id 1ltUcq-0000c5-6V; Wed, 16 Jun 2021 08:26:34 -0400
Received: by made-up.example.com with SMTP id f10so2802439iok.6;
 Wed, 16 Jun 2021 05:26:30 -0700 (PDT)
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=1e100.net; s=20161025;
 h=x-gm-message-state:mime-version:references:in-reply-to:from:date
 :message-id:subject:to:cc;
 bh=qjUjwr8ns7lWG4Dz0HG/a3G0XfCbS8LQk2C/Dha2SNQ=;
 b=fUEi7XQ5a6EXFR0Y51gNjgb1eN8v50M+ake70y14fu3W6LqZgt/SiFiohjzg/cQPFD
 YiebuN3OdIZph8Vok5g5weVrv6Ob6AImFwevDvkxDXVFj6pWs9YtkNLaXwoAKuILNoQk
 9W0SnBIYmKFtFywY/RC4s1tetFR3PTXX1JDMNpPV4p4mo5KoVY3kGa3zUVZ98KhDRktX
 Doozoxk9U8YOD9zbEpsWUf/QDfB+Wuwm9N/bDG0py+5ODIaJpILS3h9jbHSB/nR1p8Fp
 VoxzTl3URDV8KVGP2eHt6NvIRYGIFocgOTrYAjUhnfYrXIB+yOfcwop2mRJ00XGuuy4U
 N9Tg==
X-Gm-Message-State: AOAM531n4ebkPpVzP4xuX8cLevLloJuH+MgH4Ou+UUBklCqHyuDoqG1c
 RgYOBH6rKgjIq5zGTQ+C4PeCnuD6fXEmOeXzalJu6TX8UWbRPw==
X-Received: by 2002:a6b:287:: with SMTP id 129mr3604568ioc.182.1623846389248; 
 Wed, 16 Jun 2021 05:26:29 -0700 (PDT)
References: <1234@example.com>
X-Spam_score_int: -20
X-Spam_score: -2.1
X-Spam_bar: --
X-Spam_report: (-2.1 / 5.0 requ) BAYES_00=-1.9, DKIM_SIGNED=0.1,
 DKIM_VALID=-0.1, DKIM_VALID_AU=-0.1, DKIM_VALID_EF=-0.1, FREEMAIL_FROM=0.001,
 RCVD_IN_DNSWL_NONE=-0.0001, SPF_HELO_NONE=0.001,
 SPF_PASS=-0.001 autolearn=ham autolearn_force=no
X-Spam_action: no action
X-BeenThere: list@example.com
X-Mailman-Version: 2.1.23
Precedence: list
List-Id: <list@example.com>
List-Unsubscribe: <https://example.com/unsubscribe>
List-Post: <mailto:list@example.com>
List-Help: <mailto:list-request@example.com?subject=help>
List-Subscribe: <mailto:list-request@example.com?subject=subscribe>
X-Mimecast-Impersonation-Protect: Policy=CLT - Impersonation Protection Definition;Similar Internal Domain=false;Similar Monitored External Domain=false;Custom External Domain=false;Mimecast External Domain=false;Newly Observed Domain=false;Internal User Name=false;Custom Display Name List=false;Reply-to Address Mismatch=false;Targeted Threat Dictionary=false;Mimecast Threat Dictionary=false;Custom Threat Dictionary=false
X-Mimecast-Bulk-Signature: yes
X-Mimecast-Spam-Signature: bulk
X-Scanned-By: MIMEDefang 2.78 on 10.11.54.6
X-Mimecast-Spam-Score: 0
X-Mimecast-Originator: nongnu.org

On Fri, Jun 18, 2021 at 01:35:40PM -0500, Alice wrote:
> expected line 1
> 
> lots
> of
> random
> filler,
> all
> of
> which
> forms
> new
> content.

expected line 1

the line describing what is about to be quoted is not the same as new content.

> 
> skip-quoted
> should
> never
> jump
> to
> the
> middle
> of
> a
> long
> sequence
> of
> new
> content.
> 
> > Here we have
> > a few lines
> > of quoted
> > text
> 
> expected line 2
> 
> more
> random
> text
> 
> > just one line of quoted text
> 
> expected line 3
> 
> > Now, for a really long line
> > Suspendisse urna, enim, elementum magnis. Montes viverra ve, bibendum quisque egestas. Sollicitudin feugiat, venenatis dui feugiat a, rutrum cras, vulputate cursus sit tellus. Proin eros a lacus. Et, suspendisse urna, hymenaeos nec, eros. Cursus duis pede vel nunc eleifend. Lacus. Id aptent auctor, amet, duis conubia pretium turpis. Est et elementum euismod, aptent eros imperdiet aptent nascetur quis, per. Quam erat dolor tortor at erat et dolor urna nibh sed. Tempor felis pede a diam lorem placerat. Dictum, magnis ac primis ridiculus leo, dapibus tempus pede pede eni.
> 
> expected line 4

expected line 2

> 
> nothing to see beyond this point
```

**Comment by ebblake** [comment]
Fixed by #2983
